File: test/clitest.pl
use strict;
use warnings;

my $__CLI_TESTS_FAILED__ = 0;

sub assert_output {
    my ($string, $exp_out, $exp_stat) = @_;
    my (undef, $filename, $line) = caller;

    my $RED = "\x1B[31m";
    my $END = "\x1B[0m";

    my $output = qx+../$string+;
    my $status = $?;
    my $exitcode = !($output eq $exp_out && $status == $exp_stat) || 0;

    if ($output ne $exp_out) {
        print $RED . "output assert failed in $filename (line $line): $exp_out =/= $output" . $END . "\n";
    }
    if ($status != $exp_stat) {
        print $RED . "exitcode assert failed in $filename (line $line): $status =/= $exp_stat" . $END . "\n";
    }

    $__CLI_TESTS_FAILED__ ||= $exitcode;
    return $exitcode;
}

assert_output("YASL -V", "YASL v0.3.6\n", 0);
assert_output("YASL -h",
              "usage: yasl [option] [file]\n" .
              "options:\n" .
              "\t-h: this menu\n" .
              "\t-V: print current version\n" .
              "\tfile: name of file containing script\n",
              0);

exit $__CLI_TESTS_FAILED__;

File: test/clitest.c
#include "test/test_util.h"
#include <string.h>
#include <stdio.h>
#define PREF cli

const char *const cmd = "./yasl";

#define USAGESTR "usage: yasl [option] [input]\n"\
  "options:\n"\
  "\t-C: checks `input` for syntax errors but doesn't run it.\n"\
  "\t-e input: executes `input` as code and prints result of last statement.\n"\
  "\t-E input: executes `input` as code.\n"\
  "\t-h: show this text.\n"\
  "\t-V: print current version.\n"\
  "\tinput: name of file containing script (or literal to execute with -e or -E).\n"

#define TST __FILE__, __LINE__

static const struct {
	const char *out;
	const char *file;
	int line;
	const char *args[3];
} tests[] =
	{{"YASL v0.6.6\n", TST, {"-V", NULL}},
	 {"10\n", TST, {"-e", "let x=10; x;", NULL}},
	 {"", TST, {"-E", "let x=10; x;", NULL}},
	 {"10\n", TST, {"-E", "let x=10; echo x;", NULL}},
	 {USAGESTR, TST, {"-h", NULL}}};


int MAIN(void) {
	int fails = 0;
	for (size_t i = 0; i < sizeof(tests)/sizeof(tests[0]); i++) {
		if (assert_output(cmd, tests[i].args,
				  strlen(tests[i].out), tests[i].out,
				  0, NULL, 0,
				  tests[i].file, tests[i].line)) {
			fails++;
		}
	}
	REPORT((int)(sizeof(tests)/sizeof(tests[0])), fails);
	return !!fails;
}
